File: String/8. String to Integer (atoi).cpp
class Solution {
  public:
      int myAtoi(string s) {
          int i = 0;
          int sign = 1; // 正負號，1 表示正，-1 表示負
          long result = 0; // 使用 long 處理溢出
  
          // 跳過前面空格
          while (i < s.length() && s[i] == ' ') {
              i++;
          }
  
          // 如果已經到達字串末尾，返回 0
          if (i >= s.length()) return 0;
  
          // 檢查正負號
          if (s[i] == '-' || s[i] == '+') {
              sign = (s[i] == '-') ? -1 : 1;
              i++;
          }
  
          // 處理數字
          while (i < s.length() && isdigit(s[i])) {
              result = result * 10 + (s[i] - '0');
  
              // 檢查溢出 條件4
              if (result > INT_MAX) {
                  return (sign == 1) ? INT_MAX : INT_MIN;
              }
              i++;
          }
  
          // 應用正負號並返回
          return sign * result;
      }
  };

File: String/8. String to Integer (atoi)/8. String to Integer (atoi).py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = 0
        sign = 1
        result = 0

        INT_MAX = 2**31 - 1
        INT_MIN = -2**31

        while i < len(s) and s[i] == ' ':
            i += 1
        
        if i >= len(s):
            return 0

        if s[i] in ['-', '+']:
            sign = -1 if s[i] == '-' else 1
            i += 1
        
        while i < len(s) and s[i].isdigit():
            result = result * 10 + int(s[i])

            if result > INT_MAX:
                return INT_MAX if sign == 1 else INT_MIN

            i += 1

        return sign * result
        